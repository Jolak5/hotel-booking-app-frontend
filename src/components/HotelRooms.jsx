import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { FaFacebook, FaTwitter } from 'react-icons/fa';
import { AiOutlineMail } from 'react-icons/ai';
import { BiLeftArrow, BiRightArrow } from 'react-icons/bi';
import { Carousel } from 'react-responsive-carousel';
import 'react-responsive-carousel/lib/styles/carousel.min.css';
import { Link } from 'react-router-dom';
import { fetchhotels } from '../redux/Home/homeSlice';
import styles from '../styles/Home.module.css';

const HotelRooms = () => {
  const { fetched, isLoading, hotels } = useSelector((state) => state.home);
  const dispatch = useDispatch();

  const [itemsToShow, setItemsToShow] = useState(window.innerWidth > 768 ? 3 : 1);

  const handleResize = () => {
    setItemsToShow(window.innerWidth > 768 ? 3 : 1);
  };

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchhotels());
    }

    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [dispatch, fetched]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }

  const truncateDescription = (description, words = 4) => {
    const wordArray = description.trim().split(' ');
    const truncated = wordArray.slice(0, words).join(' ');
    return wordArray.length > words ? `${truncated}...` : truncated;
  };

  return (
    <>
      <div className={styles.homediv}>
        <h1 className={styles.hoteltitle}>Hotel Rooms</h1>
        <h3 className={styles.hotelsubtitle}>Please select a model</h3>
        <div className={styles.dotedbordercontainer} />
        <div className={styles.hoteldiv}>
          <Carousel
            showThumbs={false}
            emulateTouch
            dynamicHeight={false}
            showStatus={false}
            showIndicators={false}
            infiniteLoop
            useKeyboardArrows
            centerMode
            centerSlidePercentage={100 / itemsToShow}
            renderArrowPrev={(onClickHandler, hasPrev, label) => hasPrev && (
              <button type="button" onClick={onClickHandler} title={label} className={`${styles.carouselButton} ${styles.prev}`}>
                <BiLeftArrow className={styles.svgicon} />
              </button>
            )}
            renderArrowNext={(onClickHandler, hasNext, label) => hasNext && (
            <button type="button" onClick={onClickHandler} title={label} className={`${styles.carouselButton} ${styles.next}`}>
              <BiRightArrow className={styles.svgicon} />
            </button>
            )}
          >
            {hotels && hotels.length > 0 ? (
              hotels.map((hotel) => (
                <div key={hotel.id}>
                  <Link to={`/details/${hotel.id}`} className={styles.hotelinfo}>
                    <img className={styles.hotelimg} src={hotel.image.url} alt={hotel.name} />
                    <h3 className={styles.hotelname}>{hotel.name}</h3>
                    <div className={styles.dotedbordercontainer} />
                    <p className={styles.hotelDesc}>
                      {truncateDescription(hotel.description)}
                    </p>
                    <div className={styles.hotelicons}>
                      <FaFacebook className={styles.hotelicon} />
                      <FaTwitter className={styles.hotelicon} />
                      <AiOutlineMail className={styles.hotelicon} />
                    </div>
                  </Link>
                </div>
              ))
            ) : (
              <p>It is empty</p>
            )}
          </Carousel>
        </div>
      </div>
    </>
  );
};

export default HotelRooms;
