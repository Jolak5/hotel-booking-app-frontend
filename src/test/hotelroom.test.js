import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import HotelRooms from '../components/HotelRooms';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Home', () => {
    const initialState = {
        home: {
            fetched: true,
            isLoading: false,
            hotels: [
                {
                    "id": 1,
                    "name": "Hotel 1",
                    "description": "This is a description for Hotel 1.",
                    "duration": 4,
                    "price": "431.9",
                    "image": {
                        "url": "http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--70dd5f31c81314bc4a67fa2c2c0698786debcc95/image1.jpg"
                    }
                },
            ],
        },
    };

    const store = mockStore(initialState);

    test('renders without crashing', () => {
        const { getAllByText } = render(
            <Provider store={store}>
                <Router>
                    <HotelRooms />
                </Router>
            </Provider>,
        );
        const elements = getAllByText('Hotel 1');
        expect(elements.length).toBeGreaterThan(0);
    });


    test('matches snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <Router>
                        <HotelRooms />
                    </Router>
                </Provider>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});