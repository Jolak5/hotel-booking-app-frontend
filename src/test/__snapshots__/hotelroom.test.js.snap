// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<div
  className="homediv"
>
  <h1
    className="hoteltitle"
  >
    Hotel Rooms
  </h1>
  <h3
    className="hotelsubtitle"
  >
    Please select a model
  </h3>
  <div
    className="dotedbordercontainer"
  />
  <div
    className="hoteldiv"
  >
    <div
      className="carousel-root"
      tabIndex={0}
    >
      <div
        className="carousel carousel-slider"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="slider-wrapper axis-horizontal"
          style={Object {}}
        >
          <ul
            className="slider animated"
            style={
              Object {
                "MozTransitionDuration": "350ms",
                "OTransform": "translate3d(-7.105427357601002e-15%,0,0)",
                "OTransitionDuration": "350ms",
                "WebkitTransform": "translate3d(-7.105427357601002e-15%,0,0)",
                "WebkitTransitionDuration": "350ms",
                "msTransform": "translate3d(-7.105427357601002e-15%,0,0)",
                "msTransitionDuration": "350ms",
                "transform": "translate3d(-7.105427357601002e-15%,0,0)",
                "transitionDuration": "350ms",
              }
            }
          >
            <li
              className="slide selected previous"
              onClick={[Function]}
              style={
                Object {
                  "minWidth": "33.333333333333336%",
                }
              }
            >
              <div>
                <a
                  className="hotelinfo"
                  href="/details/1"
                  onClick={[Function]}
                >
                  <img
                    alt="Hotel 1"
                    className="hotelimg"
                    src="http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--70dd5f31c81314bc4a67fa2c2c0698786debcc95/image1.jpg"
                  />
                  <h3
                    className="hotelname"
                  >
                    Hotel 1
                  </h3>
                  <div
                    className="dotedbordercontainer"
                  />
                  <p
                    className="hotelDesc"
                  >
                    This is a description...
                  </p>
                  <div
                    className="hotelicons"
                  >
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M504 256C504 119 393 8 256 8S8 119 8 256c0 123.78 90.69 226.38 209.25 245V327.69h-63V256h63v-54.64c0-62.15 37-96.48 93.67-96.48 27.14 0 55.52 4.84 55.52 4.84v61h-31.28c-30.8 0-40.41 19.12-40.41 38.73V256h68.78l-11 71.69h-57.78V501C413.31 482.38 504 379.78 504 256z"
                      />
                    </svg>
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
                      />
                    </svg>
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 1024 1024"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                      />
                    </svg>
                  </div>
                </a>
              </div>
            </li>
            <li
              className="slide selected previous"
              onClick={[Function]}
              style={
                Object {
                  "minWidth": "33.333333333333336%",
                }
              }
            >
              <div>
                <a
                  className="hotelinfo"
                  href="/details/1"
                  onClick={[Function]}
                >
                  <img
                    alt="Hotel 1"
                    className="hotelimg"
                    src="http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--70dd5f31c81314bc4a67fa2c2c0698786debcc95/image1.jpg"
                  />
                  <h3
                    className="hotelname"
                  >
                    Hotel 1
                  </h3>
                  <div
                    className="dotedbordercontainer"
                  />
                  <p
                    className="hotelDesc"
                  >
                    This is a description...
                  </p>
                  <div
                    className="hotelicons"
                  >
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M504 256C504 119 393 8 256 8S8 119 8 256c0 123.78 90.69 226.38 209.25 245V327.69h-63V256h63v-54.64c0-62.15 37-96.48 93.67-96.48 27.14 0 55.52 4.84 55.52 4.84v61h-31.28c-30.8 0-40.41 19.12-40.41 38.73V256h68.78l-11 71.69h-57.78V501C413.31 482.38 504 379.78 504 256z"
                      />
                    </svg>
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
                      />
                    </svg>
                    <svg
                      className="hotelicon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 1024 1024"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                      />
                    </svg>
                  </div>
                </a>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
